### ISSUE TITLE
Add continuations to `vim.lsp.buf.*` functions.

### ISSUE THREAD

**Issue by lewis6991** [neovim/neovim#31206]
### Problem

# Problem

All `vim.lsp.buf.*` functions run asynchronously, however they do not provide callbacks when they finish.

This is needed so users can invoke other behaviours after calling these functions. And in general, asynchronous functions should always provide continuations of some form.

# Proposal

- Add a `callback` argument after every `opts` field in `vim.lsp.buf*`.
  - `vim.lsp.buf.hover(config)` -> `vim.lsp.buf.hover(opts, callback)`
  - `vim.lsp.buf.declaration(opts)` -> `vim.lsp.buf.declaration(opts, callback)`
  - `vim.lsp.buf.definition(opts)` -> `vim.lsp.buf.definition(opts, callback)`
  - `vim.lsp.buf.type_definition(opts)` -> `vim.lsp.buf.type_definition(opts, callback)`
  - `vim.lsp.buf.implementation(opts)` -> `vim.lsp.buf.implementation(opts, callback)`
  - `vim.lsp.buf.signature_help(config)` -> `vim.lsp.buf.signature_help(opts, callback)`
  - `vim.lsp.buf.format(opts)` -> `vim.lsp.buf.format(opts, callback)`
  - `vim.lsp.buf.rename(new_name, opts)` -> `vim.lsp.buf.rename(new_name, opts, callback)`
  - `vim.lsp.buf.references(context, opts)` -> `vim.lsp.buf.references(context, opts, callback)`
  - `vim.lsp.buf.document_symbol(opts)` -> `vim.lsp.buf.document_symbol(opts, callback)`
  - `vim.lsp.buf.incoming_calls()` -> `vim.lsp.buf.incoming_calls(callback)`
  - `vim.lsp.buf.outgoing_calls()` -> `vim.lsp.buf.outgoing_calls(callback)`
  - `vim.lsp.buf.typehierarchy(kind)` -> `vim.lsp.buf.typehierarchy(kind, callback)`
  - `vim.lsp.buf.workspace_symbol(query, opts)` -> `vim.lsp.buf.workspace_symbol(query, opts, callback)`
  - `vim.lsp.buf.document_highlight()` -> `vim.lsp.buf.document_highlight(callback)`
  - `vim.lsp.buf.code_action(opts)` -> `vim.lsp.buf.code_action(opts, callback)`
- Deprecate the `async` and `sync` options from `vim.lsp.buf.format()`
- Deprecate `vim.lsp.buf_request_sync()`
- (optional) If `callback` is not provided, then functions are run synchronously (like `vim.uv.*`).
  - How would this impact mappings?

## Notes:
- For functions that don't have an `opts`, `callback` will be the only option. If `opts` is required in the future then it can be added with overloads.
- `callback` could alternatively be made an option in `opts`, however this is likely to make it more awkward to use with what gets implemented in #19624, which will include `async.wrap(fun, cb_pos)` function.

## Resolves:
- #28140 
- #28004
- #27400
- #25259 
- #24168

### Expected behavior

-

**Comment by mfussenegger** [comment]
What also came up is having the ability to cancel the actions (https://github.com/neovim/neovim/pull/18372) (including LSP's cancel mechanism, not just client side).
I'm not sure how that could be solved with the callback approach, instead we'd probably need a model like we're using for `vim.system` - some kind of result object with `:cancel`, `:(a)wait` and so on. 
For chained actions like rename, where there is a (optional) prepare step, this would then need to cancel the current active operation & short-circuit.

**Comment by lewis6991** [comment]
Yes, I guess the only way to do that would be for the functions to return a table with a `cancel` method. Then the implementations can keep adjusting this cancel method depending on the stage it is at.

This would also rule out:

> (optional) If callback is not provided, then functions are run synchronously (like vim.uv.*).

Since the returned object can provide a `:wait` method.

**Comment by justinmk** [comment]
> This would also rule out:
> 
> > (optional) If callback is not provided, then functions are run synchronously (like vim.uv.*).

Why does that rule that out? If synchronous, the returned `:wait()` would be a no-op.

@gregorias does either of these approaches harmonize better with your proposal/design in https://github.com/neovim/neovim/issues/19624#issuecomment-2466700118  ?

**Comment by lewis6991** [comment]
> Why does that rule that out? If synchronous, the returned :wait() would be a no-op.

I made the same design decision with `vim.system`. if you want it to be synchronous you just suffix the call with `:wait()`.  As opposed to the opposite of adding an empty callback to get async.

I think async should be default, like it is now, as using continuations is an uncommon use case. Most people just want to call these functions in mappings.

**Comment by gregorias** [comment]
## Compatibility with asynchrony

> @gregorias does either of these approaches harmonize better with your proposal/design in https://github.com/neovim/neovim/issues/19624#issuecomment-2466700118 , or is it orthogonal ?

There must be a way to **provide a callback that is called asynchronously by the server** (or whatever handles executing the long-running operation). Other than that, whether this is done with a callback in the first call or through some returned future/promise object doesn’t matter for achieving asynchronous semantics:

```lua
-- This…
vim.lsp.buf.code_action(foo, function(result) coroutine.resume(this_coroutine, result) end)

-- … or this are fine.
local ca_promise = vim.lsp.buf.code_action(foo)
-- Register a callback to be executed upon completion.
ca_promise:await(function(result) coroutine.resume(this_coroutine, result) end)
```

What must not happen is to only have a `Future:wait` that blocks until completion in the original thread. There must be something that resumes the continuation while allowing the original thread to continue.

Theoretically, you could turn `Future:wait` into a coroutine, but then you’d need to think about what the protocol for your coroutines is and you end up implementing https://github.com/neovim/neovim/issues/19624, so I think it’s out of scope here. Better stick with a callback somewhere.

## Additional design notes

Please take these as just sharing thoughts for your consideration.

### On consistency

> I made the same design decision with vim.system. if you want it to be synchronous you just suffix the call with :wait(). As opposed to the opposite of adding an empty callback to get async.

On its own this sounds reasonable but consider also consistency with the rest of Neovim’s API. It adds mental load for `vim.uv` and `vim.lsp` to use different call semantics. For one module to use continuation callbacks as a function parameter and for the other to return promise-like objects sounds like quite a mixing of approaches.

### CancellationToken

You can still have cancellation without returning a promise-like object. You can pass something like C#’s cancellation token: an observable object that the callee can register observers (callbacks) on whenever the caller requests cancellation. Here’s a pseudo-code for how this could look like:

```lua
function vim.lsp.foo(args, cancellation_token, callback)
  local request_id = start_foo(callback)
  cancellation_token:register(function()
    send_cancellation_request(request_id)
  end)
end
```

```lua
-- caller
vim.lsp.foo(args, cancellation_token, on_done)
-- …
-- some time later
cancellation_token:cancel()
```

You would probably want to keep `cancellation_token` and `callback` inside a single table parameter. I would also expect the `token` to have an `unregister` option.

### Callback in opts

> callback could alternatively be made an option in opts, however this is likely to make it more awkward to use with what gets implemented in https://github.com/neovim/neovim/issues/19624,

I don’t think it matters much. It will require a bit of new plumbing, but it will mostly stay hidden from end-users. However, having a callback as a separate parameter does feel like the most common approach.

**Comment by lewis6991** [comment]
There's active work on this so moving to the 0.12 milestone. Can push out if the concurrency stuff also doesn't make 0.12.
